test/unit/org/apache/cassandra/hints/HintsUpgradeTest.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.cassandra.io.util.FileInputStreamPlus;
+import org.apache.cassandra.schema.KeyspaceParams;
+import org.apache.cassandra.schema.SchemaTestUtil;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class HintsUpgradeTest
+{
+    static
+    {
+        DatabaseDescriptor.daemonInitialization();
+    }
+
+    private static final String KEYSPACE = "Keyspace1";
+    private static final String TABLE = "Standard1";
+    private static final String CELLNAME = "name";
+
+    private static final String DATA_DIR = "test/data/legacy-hints/";
+    private static final String PROPERTIES_FILE = "hash.txt";
+    private static final String HOST_ID_PROPERTY = "hostId";
+    private static final String CFID_PROPERTY = "cfid";
+    private static final String CELLS_PROPERTY = "cells";
+    private static final String DESCRIPTOR_TIMESTAMP_PROPERTY = "descriptorTimestamp";
+    private static final String HASH_PROPERTY = "hash";
+
+    static TableMetadata.Builder metadataBuilder = TableMetadata.builder(KEYSPACE, TABLE)
+                                                                .addPartitionKeyColumn("key", AsciiType.instance)
+                                                                .addClusteringColumn("col", AsciiType.instance)
+                                                                .addRegularColumn("val", BytesType.instance)
+                                                                .addRegularColumn("val0", BytesType.instance)
+                                                                .compression(SchemaLoader.getCompressionParameters());
+
+    @BeforeClass
+    public static void initialize()
+    {
+        SchemaLoader.loadSchema();
+        SchemaLoader.createKeyspace(KEYSPACE, KeyspaceParams.simple(1));
+    }
+
+    @After
+    public void teardown()
+    {
+        SchemaTestUtil.announceTableDrop(KEYSPACE, TABLE);
+    }
+
+    private static class HintProperties
+    {
+        private UUID hostId;
+        private String tableId;
+        private int cells;
+        private long descriptorTimestamp;
+        private int hash;
+        private File dir;
+    }
+
+    private HintProperties init(String version) throws Exception
+    {
+        HintProperties properties = loadHintProperties(DATA_DIR + version);
+        SchemaTestUtil.announceNewTable(metadataBuilder.id(TableId.fromString(properties.tableId)).build());
+        return properties;
+    }
+
+    @Test // version 1 of hints
+    public void test30() throws Exception
+    {
+        HintProperties properties = init("3.0.29");
+        readHints(properties);
+    }
+
+    @Test // version 2 of hints
+    public void test41() throws Exception
+    {
+        HintProperties properties = init("4.1.3");
+        readHints(properties);
+    }
```

Comment:
nit: how can we fail when new hints versions are added but there's no corresponding test…?